<a href="https://colab.research.google.com/github/HIDAN122/HIDAN122/blob/main/lab_7.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

**Лабораторна робота №7**

**Тема:** “Реалізація регресійної моделі на основі нейронної мережі”

**Мета:** Нехай ми маємо набір даних вартості будинків. Нам потрібно реалізувати модель
для передбачення вартості будинку в залежності від характеристик будинку. Модель
має бути реалізована з використанням Keras та TensorFlow.

Імпортую бібліотеки

In [ ]:
import sklearn
import numpy as np
import sklearn.datasets as datasets
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.pipeline import Pipeline
from sklearn.linear_model import Ridge
from sklearn.preprocessing import StandardScaler, PolynomialFeatures
from sklearn.metrics import mean_squared_error
import math

Отримання набору даних

In [ ]:
houses = datasets.fetch_california_housing()

Перемішуємо дані перед тим, як розділити

In [ ]:
houses.data.shape

(20640, 8)

In [ ]:
houses.target.shape

(20640,)

Зміщую набір даних, щоб уникнути розбіжностей між тренувальним та тестовим набором даних

In [ ]:
permutation = np.random.permutation(houses.data.shape[0])
houses.data[permutation] = houses.data
houses.target[permutation] = houses.target

Розділяємо набір даних

In [ ]:
train_inputs, rest_inputs, train_labels, rest_labels = train_test_split(houses.data, houses.target, test_size = 0.30)
valid_inputs, tests_inputs, valid_labels, tests_labels = train_test_split(rest_inputs, rest_labels, test_size = 0.5)

Лінйна регресія

медіана та стандартне відхилення набору даних

In [ ]:
np.mean(train_inputs[:, 0]), np.std(train_inputs[:, 0])

(3.8809336240310084, 1.9076150735956234)

характеристика даного набору даних


In [ ]:
houses.feature_names

['MedInc',
 'HouseAge',
 'AveRooms',
 'AveBedrms',
 'Population',
 'AveOccup',
 'Latitude',
 'Longitude']

Модель для передбачення вартості будинку

regressor - підгонка даних до рядка

In [ ]:
model = Pipeline([
    ('normalizer', StandardScaler()),
    ('poly-feat', PolynomialFeatures(degree=2)),
    ('regressor', Ridge())
]).fit(train_inputs, train_labels)

Розраховуємо похибку

In [ ]:
squared_error = mean_squared_error(tests_labels, 
                                   model.predict(tests_inputs))
squared_error

0.4603945948361715

Перетворюю похибку в кількість в доларах 

In [ ]:
int(math.exp(squared_error) * 1000)

1584

**Висновок:** Отже, середня помилка між фактичними і прогнозованими значеннями становить 1584 долари. А на даній лабораторній роботі реалізував модель для передбачення вартості будинку в залежності від характеристик будинку